// extensions-modules/lib/configurations/config.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the Source EULA. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require('fs');
const path = require("path");
const SharedConstants = require("../models/constants");
/*
* Config class handles getting values from config.json.
*/
class Config {
    constructor(extensionConfigSectionName, fromBuild) {
        this._configJsonContent = undefined;
        this._extensionConfigSectionName = undefined;
        this._fromBuild = undefined;
        this._extensionConfigSectionName = extensionConfigSectionName;
        this._fromBuild = fromBuild;
    }
    get configJsonContent() {
        if (this._configJsonContent === undefined) {
            this._configJsonContent = this.loadConfig();
        }
        return this._configJsonContent;
    }
    getDownloadUrl() {
        return this.getConfigValue(SharedConstants.downloadUrlConfigKey);
    }
    getInstallDirectory() {
        return this.getConfigValue(SharedConstants.installDirConfigKey);
    }
    getExecutableFiles() {
        return this.getConfigValue(SharedConstants.executableFilesConfigKey);
    }
    getPackageVersion() {
        return this.getConfigValue(SharedConstants.versionConfigKey);
    }
    getConfigValue(configKey) {
        let json = this.configJsonContent;
        let toolsConfig = json[SharedConstants.serviceConfigKey];
        let configValue = undefined;
        if (toolsConfig !== undefined) {
            configValue = toolsConfig[configKey];
        }
        return configValue;
    }
    getExtensionConfig(key, defaultValue) {
        let json = this.configJsonContent;
        let extensionConfig = json[this._extensionConfigSectionName];
        let configValue = extensionConfig[key];
        if (!configValue) {
            configValue = defaultValue;
        }
        return configValue;
    }
    getWorkspaceConfig(key, defaultValue) {
        let json = this.configJsonContent;
        let configValue = json[key];
        if (!configValue) {
            configValue = defaultValue;
        }
        return configValue;
    }
    loadConfig() {
        let configContent = undefined;
        if (this._fromBuild) {
            let remainingPath = '../../../../../extensions/' + this._extensionConfigSectionName + '/client/out/config.json';
            configContent = fs.readFileSync(path.join(__dirname, remainingPath));
        }
        else {
            configContent = fs.readFileSync(path.join(__dirname, '../../../../client/out/config.json'));
        }
        return JSON.parse(configContent);
    }
}
exports.default = Config;
