// extensions-modules/lib/models/constants.js
"use strict";
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the Source EULA. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/
Object.defineProperty(exports, "__esModule", { value: true });
//constants
exports.configLogDebugInfo = 'logDebugInfo';
exports.serviceNotCompatibleError = "Client is not compatible with the service layer";
exports.serviceDownloading = "Downloading";
exports.serviceInstalling = "Installing";
exports.unsupportedPlatformErrorMessage = "This platform is unsupported and application services may not function correctly";
exports.extensionActivated = 'activated.';
exports.extensionDeactivated = 'de-activated.';
exports.configEnabled = 'enabled';
exports.configUseDebugSource = 'useDebugSource';
exports.serviceConfigKey = 'service';
exports.executableFilesConfigKey = 'executableFiles';
exports.versionConfigKey = 'version';
exports.downloadUrlConfigKey = 'downloadUrl';
exports.installDirConfigKey = 'installDir';
exports.serviceCrashButton = "View Known Issues";
exports.configDebugSourcePath = 'debugSourcePath';
exports.neverShowAgain = "Don't show again";
exports.ignorePlatformWarning = 'ignorePlatformWarning';
exports.usingDefaultPlatformMessage = "Unknown platform detected, defaulting to Linux_x64 platform";
exports.serverConnectionMetadata = "serverConnectionMetadata";
